SelfAdvancingTimerImplTest flakes when test machine is under load

SelfAdvancingTimerImplTest flakes in pipelines sometimes with:

```
io.temporal.internal.testservice.SelfAdvancingTimerImplTest > testOrdering FAILED
    java.lang.AssertionError: expected:<20000.0> but was:<20055.0>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:555)
        at org.junit.Assert.assertEquals(Assert.java:685)
        at io.temporal.internal.testservice.SelfAdvancingTimerImplTest.testOrdering(SelfAdvancingTimerImplTest.java:82)
```

--- a/temporal-sdk/src/test/java/io/temporal/internal/testservice/SelfAdvancingTimerImplTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/internal/testservice/SelfAdvancingTimerImplTest.java
@@ -21,7 +21,10 @@ package io.temporal.internal.testservice;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
+import java.time.Clock;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -35,51 +38,60 @@ import org.junit.Test;
 
 public class SelfAdvancingTimerImplTest {
 
-  public static final int INITIAL_TIME = 1;
-  private SelfAdvancingTimer timer;
-  private LongSupplier clock;
+  private static final int INITIAL_TIME = 1;
+  private static long INITIAL_SYSTEM_TIME;
+
+  private Clock mockedSystemClock;
+  private SelfAdvancingTimer fixedTimer;
+  private LongSupplier timerBasedClock;
 
   @Before
   public void setUp() throws Exception {
-    timer = new SelfAdvancingTimerImpl(INITIAL_TIME);
-    clock = timer.getClock();
+    INITIAL_SYSTEM_TIME = System.currentTimeMillis();
+    this.mockedSystemClock = mock(Clock.class);
+    when(mockedSystemClock.millis()).thenReturn(INITIAL_SYSTEM_TIME);
+    this.fixedTimer = new SelfAdvancingTimerImpl(INITIAL_TIME, mockedSystemClock);
+    this.timerBasedClock = fixedTimer.getClock();
   }
 
   @After
   public void tearDown() throws Exception {
-    timer.shutdown();
+    fixedTimer.shutdown();
   }
 
   @Test
   public void testSchedule() throws InterruptedException {
     AtomicLong captured = new AtomicLong();
-    timer.schedule(Duration.ofDays(1), () -> captured.set(clock.getAsLong()));
+    fixedTimer.schedule(Duration.ofDays(1), () -> captured.set(timerBasedClock.getAsLong()));
     Thread.sleep(100);
     assertTrue(Duration.ofDays(1).toMillis() + INITIAL_TIME <= captured.get());
-    assertTrue(Duration.ofDays(1).toMillis() + INITIAL_TIME <= clock.getAsLong());
-    long start = clock.getAsLong();
-    timer.schedule(Duration.ofSeconds(123), () -> captured.set(clock.getAsLong()));
-    assertTrue(clock.getAsLong() - start >= Duration.ofSeconds(123).toMillis());
+    assertTrue(Duration.ofDays(1).toMillis() + INITIAL_TIME <= timerBasedClock.getAsLong());
+    long start = timerBasedClock.getAsLong();
+    fixedTimer.schedule(Duration.ofSeconds(123), () -> captured.set(timerBasedClock.getAsLong()));
+    assertTrue(timerBasedClock.getAsLong() - start >= Duration.ofSeconds(123).toMillis());
   }
 
   @Test
   public void testOrdering() throws InterruptedException {
     List<Long> captured = Collections.synchronizedList(new ArrayList<>());
-    timer.lockTimeSkipping("unit test");
-    timer.schedule(Duration.ofSeconds(100), () -> captured.add(clock.getAsLong()));
-    timer.schedule(Duration.ofSeconds(20), () -> captured.add(clock.getAsLong()));
-    timer.schedule(Duration.ofSeconds(10), () -> captured.add(clock.getAsLong()));
-    timer.schedule(Duration.ofSeconds(1), () -> captured.add(clock.getAsLong()));
-    timer.unlockTimeSkipping("unit test");
+    fixedTimer.lockTimeSkipping("unit test");
+    fixedTimer.schedule(Duration.ofSeconds(100), () -> captured.add(timerBasedClock.getAsLong()));
+    when(mockedSystemClock.millis()).thenReturn(INITIAL_SYSTEM_TIME + 5);
+    fixedTimer.schedule(Duration.ofSeconds(20), () -> captured.add(timerBasedClock.getAsLong()));
+    when(mockedSystemClock.millis()).thenReturn(INITIAL_SYSTEM_TIME + 10);
+    fixedTimer.schedule(Duration.ofSeconds(10), () -> captured.add(timerBasedClock.getAsLong()));
+    when(mockedSystemClock.millis()).thenReturn(INITIAL_SYSTEM_TIME + 15);
+    fixedTimer.schedule(Duration.ofSeconds(1), () -> captured.add(timerBasedClock.getAsLong()));
+    fixedTimer.unlockTimeSkipping("unit test");
     Thread.sleep(100);
     List<Long> expected =
         Arrays.asList(
-            Duration.ofSeconds(1).toMillis(),
-            Duration.ofSeconds(10).toMillis(),
-            Duration.ofSeconds(20).toMillis(),
-            Duration.ofSeconds(100).toMillis());
+            INITIAL_TIME + Duration.ofSeconds(1).toMillis() + 15,
+            INITIAL_TIME + Duration.ofSeconds(10).toMillis() + 10,
+            INITIAL_TIME + Duration.ofSeconds(20).toMillis() + 5,
+            INITIAL_TIME + Duration.ofSeconds(100).toMillis());
     for (int i = 0; i < captured.size(); i++) {
-      assertEquals(expected.get(i), captured.get(i), 50.0);
+      assertEquals(expected.get(i), captured.get(i));
     }
   }
 }

--- a/temporal-testing/src/main/java/io/temporal/internal/testservice/SelfAdvancingTimerImpl.java
+++ b/temporal-testing/src/main/java/io/temporal/internal/testservice/SelfAdvancingTimerImpl.java
@@ -22,6 +22,7 @@ package io.temporal.internal.testservice;
 import com.google.common.util.concurrent.Uninterruptibles;
 import io.temporal.workflow.Functions;
 import java.sql.Timestamp;
+import java.time.Clock;
 import java.time.Duration;
 import java.util.*;
 import java.util.concurrent.LinkedBlockingDeque;
@@ -153,15 +154,15 @@ final class SelfAdvancingTimerImpl implements SelfAdvancingTimer {
     }
   }
 
-  private static class LockEvent {
-    String caller;
-    LockEventType lockType;
-    long timestamp;
+  private class LockEvent {
+    final String caller;
+    final LockEventType lockType;
+    final long timestamp;
 
     public LockEvent(String caller, LockEventType lockType) {
       this.caller = caller;
       this.lockType = lockType;
-      timestamp = System.currentTimeMillis();
+      this.timestamp = systemClock.millis();
     }
   }
 
@@ -209,6 +210,7 @@ final class SelfAdvancingTimerImpl implements SelfAdvancingTimer {
   }
 
   private static final Logger log = LoggerFactory.getLogger(SelfAdvancingTimerImpl.class);
+  private final Clock systemClock;
   private final LongSupplier clock =
       () -> {
         long timeMillis = this.currentTimeMillis();
@@ -235,8 +237,9 @@ final class SelfAdvancingTimerImpl implements SelfAdvancingTimer {
   private final Thread timerPump = new Thread(new TimerPump(), "SelfAdvancingTimer Pump");
   private LockHandle timeLockOnEmptyQueueHandle;
 
-  public SelfAdvancingTimerImpl(long initialTime) {
-    currentTime = initialTime == 0 ? System.currentTimeMillis() : initialTime;
+  public SelfAdvancingTimerImpl(long initialTime, Clock systemClock) {
+    this.systemClock = systemClock;
+    currentTime = initialTime == 0 ? systemClock.millis() : initialTime;
     executor.setRejectedExecutionHandler(new CallerRunsPolicy());
     // Queue is initially empty. The code assumes that in this case skipping is already locked.
     timeLockOnEmptyQueueHandle = lockTimeSkipping("SelfAdvancingTimerImpl constructor empty-queue");
@@ -248,7 +251,7 @@ final class SelfAdvancingTimerImpl implements SelfAdvancingTimer {
       if (timeLastLocked < 0 || systemTimeLastLocked < 0) {
         throw new IllegalStateException("Invalid timeLastLocked or systemTimeLastLocked");
       }
-      currentTime = timeLastLocked + (System.currentTimeMillis() - systemTimeLastLocked);
+      currentTime = timeLastLocked + (systemClock.millis() - systemTimeLastLocked);
     } else {
       TimerTask task = tasks.peek();
       if (task != null && !task.isCanceled() && task.getExecutionTime() > currentTime) {
@@ -280,6 +283,9 @@ final class SelfAdvancingTimerImpl implements SelfAdvancingTimer {
     Functions.Proc cancellationHandle;
     lock.lock();
     try {
+      // get a last time in case the last TimePump update was relatively long time ago
+      updateTimeLocked();
+
       long executionTime = delay.toMillis() + currentTime;
       TimerTask timerTask = new TimerTask(executionTime, task, taskInfo);
       cancellationHandle = () -> timerTask.cancel();
@@ -321,7 +327,7 @@ final class SelfAdvancingTimerImpl implements SelfAdvancingTimer {
   private LockHandle lockTimeSkippingLocked(String caller) {
     if (lockCount++ == 0) {
       timeLastLocked = currentTime;
-      systemTimeLastLocked = System.currentTimeMillis();
+      systemTimeLastLocked = systemClock.millis();
     }
     LockEvent event = new LockEvent(caller, LockEventType.LOCK);
     lockEvents.add(event);

--- a/temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowStoreImpl.java
+++ b/temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowStoreImpl.java
@@ -42,6 +42,7 @@ import io.temporal.failure.ApplicationFailure;
 import io.temporal.internal.common.WorkflowExecutionUtils;
 import io.temporal.internal.testservice.RequestContext.Timer;
 import io.temporal.workflow.Functions;
+import java.time.Clock;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -177,7 +178,7 @@ class TestWorkflowStoreImpl implements TestWorkflowStore {
   }
 
   public TestWorkflowStoreImpl(long initialTimeMillis) {
-    timerService = new SelfAdvancingTimerImpl(initialTimeMillis);
+    timerService = new SelfAdvancingTimerImpl(initialTimeMillis, Clock.systemDefaultZone());
     // locked until the first save
     emptyHistoryLockHandle = timerService.lockTimeSkipping("TestWorkflowStoreImpl constructor");
   }

